crypto/init.c

```diff
@@ -474,12 +474,12 @@ void OPENSSL_cleanup(void)
                     "RAND_cleanup()\n");
 
 #endif
-    CRYPTO_cleanup_all_ex_data();
-    EVP_cleanup();
-    CONF_modules_free();
 #ifndef OPENSSL_NO_ENGINE
     ENGINE_cleanup();
 #endif
+    CRYPTO_cleanup_all_ex_data();
+    EVP_cleanup();
+    CONF_modules_free();
```

Comment:
Gah! I just moved CONF_modules_free() to before ENGINE_cleanup() 773fd0bad because the former can call ENGINE code. If ENGINE_cleanup() is already called then the locks have been freed, and CONF_modules_free() can hang!